### PR TITLE
Additional work on storage and filesystem operations

### PR DESCRIPTION
**Pull request by slouken**
@flibitijibibo, @icculus, I'm jamming on the work you've been doing, adding filesystem operations to storage.

@icculus, I've renamed some of the functions in https://github.com/libsdl-org/SDL/pull/9149 to better reflect they work on both files and directories.

Feedback is welcome!

### CODE REVIEW

File: src/file/SDL_iostream.c [hunk 1]
```diff
@@ -741,12 +741,16 @@ SDL_IOStatus SDL_GetIOStatus(SDL_IOStream *context)
...
+SDL_IOStream *SDL_OpenIO(const SDL_IOStreamInterface *iface, int version, void *userdata)
 {
     if (!iface) {
         SDL_InvalidParamError("iface");
         return NULL;
     }
+    if (version != SDL_IOSTREAM_INTERFACE_VERSION) {
+        SDL_SetError("Unknown interface version");
+        return NULL;
+    }
```
Comment (sezero) [inline src/file/SDL_iostream.c:753]: Maybe make this a switch statement to imply your _possible_ intent to handle things differently with multiple known versions in future ??

Comment (slouken) [reply src/file/SDL_iostream.c:753]: Sure, that sounds good.

File: include/SDL3/SDL_filesystem.h [hunk 2]
```diff
@@ -236,6 +236,102 @@ typedef enum
...
+/**
+ * Create a directory.
+ *
+ * \param path the path of the directory to create
+ * \returns 0 on success or a negative error code on failure; call
+ *          SDL_GetError() for more information.
+ *
+ * \since This function is available since SDL 3.0.0.
+ */
+extern DECLSPEC int SDLCALL SDL_CreateDirectory(const char *path);
```
Comment (flibitijibibo) [inline include/SDL3/SDL_filesystem.h:270]: Will this be recursive by chance? Not a dealbreaker but would be cool to have SDL handle the loop here!

File: include/SDL3/SDL_storage.h [hunk 3]
```diff
@@ -100,13 +106,31 @@ extern DECLSPEC SDL_Storage *SDLCALL SDL_OpenTitleStorage(const char *override,
...
- * \sa SDL_StorageSpaceRemaining
+ * \sa SDL_GetStorageFileSize
+ * \sa SDL_ReadStorageFile
+ * \sa SDL_WriteStorageFile
+ * \sa SDL_GetStorageSpaceRemaining
  */
 extern DECLSPEC SDL_Storage *SDLCALL SDL_OpenUserStorage(const char *org, const char *app, SDL_PropertiesID props);
 
+/**
+ * Opens up a container for local filesystem storage.
```
Comment (flibitijibibo) [inline include/SDL3/SDL_storage.h:117]: Definitely worth putting a giant warning that this should be for internal development only; the worst case scenario is that the engine gets coupled to this API for release builds, meaning they'll eventually have to port from SDL to... well, SDL.

**Review by flibitijibibo — COMMENTED** [review]
Looks good so far, just a couple bits that are mostly about documentation.

File: include/SDL3/SDL_filesystem.h (hunk 2 above)
Comment (slouken) [reply include/SDL3/SDL_filesystem.h:270]: Yeah, I'm throwing a TODO in SDL_CreateDirectory()

File: include/SDL3/SDL_storage.h (hunk 3 above)
Comment (slouken) [reply include/SDL3/SDL_storage.h:136]: Do you want to draft a warning?

Comment (slouken) [reply include/SDL3/SDL_storage.h:136]: Maybe something like:
"This is provided for development and tools. Portable applications should use SDL_OpenTitleStorage() for access to game data and SDL_OpenUserStorage() for user data."

Comment (flibitijibibo) [reply include/SDL3/SDL_storage.h:136]: That works for me!

File: src/filesystem/SDL_sysfilesystem.h [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+*/
+
+#ifndef SDL_sysfilesystem_h_
+#define SDL_sysfilesystem_h_
+
+int SDL_SYS_FSenumerate(const char *fullpath, const char *dirname, SDL_EnumerateDirectoryCallback cb, void *userdata);
+int SDL_SYS_FSremove(const char *fullpath);
+int SDL_SYS_FSrename(const char *oldfullpath, const char *newfullpath);
+int SDL_SYS_FSmkdir(const char *fullpath);
+int SDL_SYS_FSstat(const char *fullpath, SDL_PathInfo *info);
```
[the remark this answers is not in the document]
Comment (slouken) [reply src/filesystem/SDL_sysfilesystem.h:29]: Yep, do you want to do that cleanup once I check in?

[the remark this answers is not in the document]
Comment (icculus) [reply src/filesystem/SDL_sysfilesystem.h:29]: Sure!

**Comment by madebr** [comment]
Depending on how long this pr is wip, would it be possible to fix the Xcode project on main?
Building the SDL3.dmg currently fails with the following link error:
```
2024-03-17T03:33:27.4029610Z Undefined symbols for architecture x86_64:
2024-03-17T03:33:27.4030240Z   "_SDL_CloseStorage_REAL", referenced from:
2024-03-17T03:33:27.4030850Z       _initialize_jumptable in SDL_dynapi.o
2024-03-17T03:33:27.4031440Z   "_SDL_OpenStorage_REAL", referenced from:
2024-03-17T03:33:27.4032020Z       _initialize_jumptable in SDL_dynapi.o
2024-03-17T03:33:27.4032640Z   "_SDL_OpenTitleStorage_REAL", referenced from:
2024-03-17T03:33:27.4033250Z       _initialize_jumptable in SDL_dynapi.o
2024-03-17T03:33:27.4034010Z   "_SDL_OpenUserStorage_REAL", referenced from:
2024-03-17T03:33:27.4034650Z       _initialize_jumptable in SDL_dynapi.o
2024-03-17T03:33:27.4035290Z   "_SDL_StorageFileSize_REAL", referenced from:
2024-03-17T03:33:27.4035940Z       _initialize_jumptable in SDL_dynapi.o
2024-03-17T03:33:27.4036570Z   "_SDL_StorageReadFile_REAL", referenced from:
2024-03-17T03:33:27.4037210Z       _initialize_jumptable in SDL_dynapi.o
2024-03-17T03:33:27.4037840Z   "_SDL_StorageReady_REAL", referenced from:
2024-03-17T03:33:27.4038700Z       _initialize_jumptable in SDL_dynapi.o
2024-03-17T03:33:27.4039410Z   "_SDL_StorageSpaceRemaining_REAL", referenced from:
2024-03-17T03:33:27.4040090Z       _initialize_jumptable in SDL_dynapi.o
2024-03-17T03:33:27.4040750Z ld: symbol(s) not found for architecture x86_64
2024-03-17T03:33:27.4041940Z clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
There might be build errors for other architectures.

**Comment by slouken** [comment]
> Depending on how long this pr is wip, would it be possible to fix the Xcode project on main? Building the SDL3.dmg currently fails with the following link error:
> 
> ```
> 2024-03-17T03:33:27.4029610Z Undefined symbols for architecture x86_64:
> 2024-03-17T03:33:27.4030240Z   "_SDL_CloseStorage_REAL", referenced from:
> 2024-03-17T03:33:27.4030850Z       _initialize_jumptable in SDL_dynapi.o
> 2024-03-17T03:33:27.4031440Z   "_SDL_OpenStorage_REAL", referenced from:
> 2024-03-17T03:33:27.4032020Z       _initialize_jumptable in SDL_dynapi.o
> 2024-03-17T03:33:27.4032640Z   "_SDL_OpenTitleStorage_REAL", referenced from:
> 2024-03-17T03:33:27.4033250Z       _initialize_jumptable in SDL_dynapi.o
> 2024-03-17T03:33:27.4034010Z   "_SDL_OpenUserStorage_REAL", referenced from:
> 2024-03-17T03:33:27.4034650Z       _initialize_jumptable in SDL_dynapi.o
> 2024-03-17T03:33:27.4035290Z   "_SDL_StorageFileSize_REAL", referenced from:
> 2024-03-17T03:33:27.4035940Z       _initialize_jumptable in SDL_dynapi.o
> 2024-03-17T03:33:27.4036570Z   "_SDL_StorageReadFile_REAL", referenced from:
> 2024-03-17T03:33:27.4037210Z       _initialize_jumptable in SDL_dynapi.o
> 2024-03-17T03:33:27.4037840Z   "_SDL_StorageReady_REAL", referenced from:
> 2024-03-17T03:33:27.4038700Z       _initialize_jumptable in SDL_dynapi.o
> 2024-03-17T03:33:27.4039410Z   "_SDL_StorageSpaceRemaining_REAL", referenced from:
> 2024-03-17T03:33:27.4040090Z       _initialize_jumptable in SDL_dynapi.o
> 2024-03-17T03:33:27.4040750Z ld: symbol(s) not found for architecture x86_64
> 2024-03-17T03:33:27.4041940Z clang: error: linker command failed with exit code 1 (use -v to see invocation)
> ```
> 
> There might be build errors for other architectures.

I’ll probably check this in tomorrow morning, but I won’t have access to a Mac until Monday or Tuesday. It should be an easy fix then.

File: docs/README-migration.md [hunk 5]
```diff
@@ -1299,7 +1299,7 @@ SDL_IOStream *SDL_RWFromFP(FILE *fp, SDL_bool autoclose)
     rwopsdata->fp = fp;
     rwopsdata->autoclose = autoclose;
 
-    rwops = SDL_OpenIO(&iface, rwopsdata);
+    rwops = SDL_OpenIO(&iface, SDL_IOSTREAM_INTERFACE_VERSION, rwopsdata);
```
[the remark this answers is not in the document]
Comment (slouken) [reply docs/README-migration.md:1302]: Hmm, that’s a good point. I’ll back out that change.

File: src/filesystem/SDL_filesystem.c [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+{
+    /* TODO: Recursively create subdirectories */
+    if (!path) {
+        return SDL_InvalidParamError("path");
+    }
+    return SDL_SYS_FSmkdir(path);
+}
+
+int SDL_EnumerateDirectory(const char *path, SDL_EnumerateDirectoryCallback callback, void *userdata)
+{
```
[the remark this answers is not in the document]
Comment (slouken) [reply src/filesystem/SDL_filesystem.c:74]: I’m on the fence too. From an application point of view, the way we return an array of results and free it at once has been super nice to work with.

On the other hand, if the app is building a glob list of files in a very large directory it’s going to be way more efficient to do that in a callback.

I wonder if maybe we can just build optional globbing into the enumeration function and have it return results?

[the remark this answers is not in the document]
Comment (icculus) [reply src/filesystem/SDL_filesystem.c:74]: Maybe we leave the callback API but also add:

```c
char **SDL_GlobDirectory(const char *path, const char *pattern);
```

`path` is a directory, `pattern` allows '*' and '?' wildcards within that directory (NULL means "don't filter, return everything"). Returns an array that you free with a single SDL_free: all the string pointers point into memory allocated in the same block as the array itself.

Internally, this uses the callback, where it filters out things that don't match the pattern.

Makes implementations easy, gives an easy public API with an added feature of pattern matching, and leaves the original for apps that prefer to drink straight from the firehose.

[the remark this answers is not in the document]
Comment (slouken) [reply src/filesystem/SDL_filesystem.c:74]: I like it. Let's add that after this is merged.

**Pull request merged**